### ISSUE TITLE
[Feature Request] Support custom client ID

### ISSUE THREAD

**Issue by jiasli** [Azure/azure-cli#22775]
**Related command**
`az login`

**Is your feature request related to a problem? Please describe.**
Currently Azure CLI uses a fixed client ID to authenticate into AAD

https://github.com/Azure/azure-cli/blob/1d973cceb38980181eeaa45934497d2148a3e7b2/src/azure-cli-core/azure/cli/core/auth/identity.py#L22

This client ID `04b07795-8ddb-461a-bbee-02f9e1bf7b46` is a First Party Application, registered in First Party Portal.

This application is only pre-authorized for a limited set of [Microsoft Graph permissions](https://docs.microsoft.com/en-us/graph/permissions-reference). Using Azure CLI command `az rest` to call [trustFrameworkKeySet](https://docs.microsoft.com/en-us/graph/api/trustframeworkkeyset-get?view=graph-rest-beta&tabs=http) API which requires **TrustFrameworkKeySet.Read.All**, **TrustFrameworkKeySet.ReadWrite.All** will fail (https://github.com/Azure/azure-cli/issues/22755).

**Describe the solution you'd like**
Support custom client ID, so that users can create their own app in their tenant, consent to that app and let Azure CLI use that app’s client ID:

```
az login --client-id
```

**Describe alternatives you've considered**
Apply for those permissions in First Party Portal to support these APIs.

**Comment by yonzhan** [comment]
Support custom client ID

**Comment by navba-MSFT** [comment]
Removing CXP Attention label from this issue since this was created internally by SDK team

**Comment by jiasli** [comment]
We received a similar issue today that the user wants to use Azure CLI's client ID to call `b2cIdentityUserFlows` API which requires delegated permission **IdentityUserFlow.Read.All**, **IdentityUserFlow.ReadWrite.All**:

https://docs.microsoft.com/en-us/graph/api/identitycontainer-list-b2cuserflows?view=graph-rest-beta&tabs=csharp

![image](https://user-images.githubusercontent.com/4003950/190593427-edd9ab8c-cd58-43e8-90f1-1bc61752bad9.png)

but Azure CLI's **Directory.AccessAsUser.All** delegated permission is not sufficient. This causes failure:

> AADSTS65002: Consent between first party application '04b07795-8ddb-461a-bbee-02f9e1bf7b46' and first party resource '00000003-0000-0000-c000-000000000000' must be configured via preauthorization - applications owned and operated by Microsoft must get approval from the API owner before requesting tokens for that API.

**Comment by jiasli** [comment]
### Workaround: Use service principal instead of user login

1. Create your own application and service principal 
2. Give the application Application Permission **IdentityUserFlow.Read.All**, **IdentityUserFlow.ReadWrite.All** 
3. Log in to Azure CLI with that service principal: https://docs.microsoft.com/en-us/cli/azure/authenticate-azure-cli#sign-in-with-a-service-principal

**Comment by kthejoker** [comment]
What if a user doesn't have permission to create a service principal?

This exact same flow worked fine in October of 2022. A user could use a DeviceCodeCredential and then access the Databricks scope with their own tokens.

Now I have to create an SP?

**Comment by jiasli** [comment]
Azure CLI's first party application `04b07795-8ddb-461a-bbee-02f9e1bf7b46` is currently registered with below Microsoft Graph delegated permissions:

- AuditLog.Read.All
- Directory.AccessAsUser.All
- Group.ReadWrite.All
- User.ReadWrite.All
- openid

This can be observed in the access token retrieved from

```
az account get-access-token --scope https://graph.microsoft.com/.default
```

The access token contains the below `scp` claim:

```
  "scp": "AuditLog.Read.All Directory.AccessAsUser.All email Group.ReadWrite.All openid profile User.ReadWrite.All",
```

**Comment by jiasli** [comment]
> This exact same flow worked fine in October of 2022. A user could use a DeviceCodeCredential and then access the Databricks scope with their own tokens.

@kthejoker, There is no `DeviceCodeCredential` concept in Azure CLI. Are you referring to `az login --use-device-code`? If so and you are seeing failure, that indicates your tenant may now have a Conditional Access policy that blocks accessing Databricks scope with Azure CLI's client ID and device code flow.

> Now I have to create an SP?

Yes.

> What if a user doesn't have permission to create a service principal?

You need to work with your tenant admin to either

-  Unblock accessing Databricks scope with Azure CLI's client ID and device code flow
- Create a service principal

**Comment by cveld** [comment]
Today I started a project to configure PIM using terraform using the azuread provider's resource `azuread_group_role_management_policy`. This requires
`RoleManagementPolicy.ReadWrite.AzureADGroup`. I am not a great fan of your suggested workaround to introduce a service principal as we work with secret-free service principals.
It seems either the required permission scopes are to be added to the first party app registration, or the custom client id should be introduced.

**Comment by cveld** [comment]
@jiasli is there any plan to extend the permission scopes on the first party app registration `04b07795-8ddb-461a-bbee-02f9e1bf7b46`? Any reason why you are keeping this a limited range? In the meanwhile the msgraph cli app registration did get the additional scopes. Although that is what I am assuming as the msgraph cli is capable of user delegated access to PIM related operations.

**Comment by jiasli** [comment]
#30149 is another issue about Azure CLI's lack of `PrivilegedAssignmentSchedule.ReadWrite.AzureADGroup` or `PrivilegedAssignmentSchedule.Remove.AzureADGroup` delegated permission to call [Create assignmentScheduleRequest](https://learn.microsoft.com/en-us/graph/api/privilegedaccessgroup-post-assignmentschedulerequests?view=graph-rest-1.0&tabs=http) API.

**Comment by jiasli** [comment]
I retrieved an access token from [Graph Explorer](https://developer.microsoft.com/en-us/graph/graph-explorer) and decoded it at https://jwt.ms/ by clicking the `{}` button:

![Image](https://github.com/user-attachments/assets/97d9edc9-2913-48dc-97ae-ee98aab75f05)

The decoded claims contain:

```
  "app_displayname": "Graph Explorer",
  "appid": "de8bc8b5-d9f9-48b1-a8ad-b748da725064",
```

`scp` claim is

```
Calendars.ReadWrite
Contacts.ReadWrite
DeviceManagementApps.ReadWrite.All
DeviceManagementConfiguration.Read.All
DeviceManagementConfiguration.ReadWrite.All
DeviceManagementManagedDevices.PrivilegedOperations.All
DeviceManagementManagedDevices.Read.All
DeviceManagementManagedDevices.ReadWrite.All
DeviceManagementRBAC.Read.All
DeviceManagementRBAC.ReadWrite.All
DeviceManagementServiceConfig.Read.All
DeviceManagementServiceConfig.ReadWrite.All
Directory.AccessAsUser.All
Directory.ReadWrite.All
Files.ReadWrite.All
Group.ReadWrite.All
IdentityRiskEvent.Read.All
Mail.ReadWrite
MailboxSettings.ReadWrite
Notes.ReadWrite.All
openid
People.Read
Policy.Read.All
Presence.Read
Presence.Read.All
profile
Reports.Read.All
Sites.ReadWrite.All
Tasks.ReadWrite
User.Read
User.ReadBasic.All
User.ReadWrite
User.ReadWrite.All
email
```

Apparently, Graph Explorer's first party app has far more delegated permissions than Azure CLI.

**Comment by c-baumgartner** [comment]
Up voting for this is feature request! Ability to havbe a custom client-id for interactive logins would really help for some terraform providers requiring very specific delegated Graph permissions along with azurerm etc.

**Comment by lthousechris** [comment]
We are encountering the same issue when moving our assignment and configuration of PIM role management policy to Terraform.  
Granting the permissions to the service principal for the managed identity running the pipeline works fine, but our users are unable to run terraform plan as their own user accounts.

**Comment by moredatapls** [comment]
I actually managed to get it to work with the built-in Azure CLI application by extending its delegated permission grants. I don't have the exact steps anymore, but it goes something like this:

1. Get the App Admin role in Entra so you can create SPs and Entra role assignments
2. If the "Microsoft Azure CLI" SP does not exist in your tenant, register it using its ID: `az ad sp create --id 04b07795-8ddb-461a-bbee-02f9e1bf7b46`
3. Figure out which permissions you need, e.g. Graph `RoleManagementPolicy.ReadWrite.AzureADGroup`, and locate it's **delegated** permission ID [here](https://learn.microsoft.com/en-us/graph/permissions-reference), e.g. `0da165c7-3f15-4236-b733-c0b0f6abe41d`
4. Create a delegated permission role assignment for the Azure CLI SP: `az ad app permission add --id 04b07795-8ddb-461a-bbee-02f9e1bf7b46 --api {api application ID} --api-permissions {delegated permission id}=Scope`, so e.g. for MS Graph and the permission above: `az ad app permission add --id 04b07795-8ddb-461a-bbee-02f9e1bf7b46 --api 00000003-0000-0000-c000-000000000000 --api-permissions 0da165c7-3f15-4236-b733-c0b0f6abe41d=Scope`
5. The delegated permission grant should now allow your users to use the permission via the Azure CLI, if they have been granted it themselves

I definitely got it to work a while back, but I give no guarantee that the steps above are correct. Maybe one of you can try out and share the exact steps?

**Comment by c-baumgartner** [comment]
@moredatapls I have tried your steps, but in **step 4** the Azure CLI throws the following error on me:

```bash
$ az ad app permission add --id 04b07795-8ddb-461a-bbee-02f9e1bf7b46 --api 00000003-0000-0000-c000-000000000000 --api-permissions 0da165c7-3f15-4236-b733-c0b0f6abe41d=Scope

Resource '04b07795-8ddb-461a-bbee-02f9e1bf7b46' does not exist or one of its queried reference-property objects are not present.
```

**Comment by moredatapls** [comment]
@c-baumgartner could you try with the Azure CLI‘s object ID instead?

**Comment by c-baumgartner** [comment]
Sorry @moredatapls, did not mentioned it.. but I have also tried using the objectId of the service principal instead of the appId - error is the same:

```bash
ID=$(az ad sp list --display-name "Microsoft Azure CLI" --query [].id -o tsv)

az ad app permission add --id $ID --api 00000003-0000-0000-c000-000000000000 --api-permissions 0da165c7-3f15-4236-b733-c0b0f6abe41d=Scope

Resource '1497399b-e7f9-458a-beb9-710b7359d027' does not exist or one of its queried reference-property objects are not present.
```

**Comment by moredatapls** [comment]
@c-baumgartner Sorry, I wrote it down from memory but used the wrong command. Managed to test it today and this works:

1. Get the App Admin role in Entra so you can create SPs and Entra role assignments
2. If the "Microsoft Azure CLI" SP does not exist in your tenant, register it using its ID: `az ad sp create --id 04b07795-8ddb-461a-bbee-02f9e1bf7b46`
3. Grant the delegated permissions that you need (see [here](https://learn.microsoft.com/en-us/cli/azure/ad/app/permission?view=azure-cli-latest#az-ad-app-permission-grant) for more details): `az ad app permission grant --api 00000003-0000-0000-c000-000000000000 --id 04b07795-8ddb-461a-bbee-02f9e1bf7b46 --scope RoleManagementPolicy.ReadWrite.AzureADGroup`

Afterwards, the delegated permission should show up in the portal and users will be able to use the relevant permission, as long as they have been granted it, too:

<img width="886" height="428" alt="Image" src="https://github.com/user-attachments/assets/2f649357-c62a-4f8a-9f07-75db450631cc" />

This obviously also works with other APIs, even internal ones.

**Comment by c-baumgartner** [comment]
Thank you @moredatapls, one step closer! But when I am trying to acquire an access token for `ms-graph`, the new Scope is not included.

I have tried to explicitly get a token for the scope, but this will throw the following message:

```bash
az account get-access-token --resource-type ms-graph --scope RoleManagementPolicy.ReadWrite.AzureADGroup

AADSTS65002: Consent between first party application '04b07795-8ddb-461a-bbee-02f9e1bf7b46' and first party resource '00000003-0000-0000-c000-000000000000' must be configured via preauthorization - applications owned and operated by Microsoft must get approval from the API owner before requesting tokens for that API
```

Also have tried to use scope `.default` instead, but only the default Azure CLI scopes for Graph are included:

```json
{
  "scp": "Application.ReadWrite.All AppRoleAssignment.ReadWrite.All AuditLog.Read.All DelegatedPermissionGrant.ReadWrite.All Directory.AccessAsUser.All email Group.ReadWrite.All openid profile User.Read.All User.ReadWrite.All"
}
```

Sorry to bother you, I am really hoping this workaround can solve a big issue while waiting for MSFT to add the requested Feature.

Regards,
Christian

**Comment by Poltergeisen** [comment]
Is there a timeline for when this will be fixed? Our terraform deployments are not working locally because of this.
